# Capstone Project - EDA: Ames Housing (House Prices)
## Phân công: Thành viên 1 - Khảo sát bài toán & xác định tính chất dữ liệu

**Nguồn dữ liệu:** Kaggle - House Prices: Advanced Regression Techniques
https://www.kaggle.com/competitions/house-prices-advanced-regression-techniques

**Paper gốc:** De Cock, D. (2011). *Ames, Iowa: Alternative to the Boston Housing Data as an End of Semester Regression Project.* Journal of Statistics Education, 19(3).
https://jse.amstat.org/v19n3/decock.pdf


## 1. Bối cảnh bài toán (theo paper De Cock, 2011)

- Dữ liệu mô tả các giao dịch mua bán nhà ở riêng lẻ tại thành phố **Ames, Iowa, Hoa Kỳ**, thu thập từ **Văn phòng Đánh giá thuế (City Assessor's Office)** trong giai đoạn **2006-2010**.
- Tác giả đề xuất bộ dữ liệu này như một **phương án thay thế cho bộ dữ liệu Boston Housing** (quá cũ, ít thuộc tính) để làm đồ án cuối kỳ môn hồi quy cho sinh viên thống kê/khoa học dữ liệu.
- Bộ dữ liệu gốc có **2930 quan sát và 82 thuộc tính**; phiên bản đưa lên cuộc thi Kaggle được chia thành **train (1460 dòng, có sẵn SalePrice)** và **test (1459 dòng, ẩn SalePrice)**, cùng **81 thuộc tính** mô tả gần như mọi khía cạnh của ngôi nhà (diện tích, chất lượng, năm xây, vị trí, tiện ích...).
- **Bài toán:** Dự đoán **SalePrice** (giá bán nhà) dựa trên các thuộc tính đầu vào -> đây là bài toán **hồi quy (regression)**.
- Lý do bộ dữ liệu này phù hợp cho EDA là có cả biến số và biến phân loại, có giá trị thiếu tự nhiên và có một số mẫu bất thường được tác giả lưu ý trong paper.


In [1]:
import pandas as pd
import numpy as np
pd.set_option('display.max_columns', 15)

df = pd.read_csv('train.csv')
print("Kích thước dữ liệu (train):", df.shape)
df.head()

Kich thuoc du lieu (train): (1460, 81)


,Id,MSSubClass,MSZoning,LotFrontage,LotArea,Street,Alley,...,MiscFeature,MiscVal,MoSold,YrSold,SaleType,SaleCondition,SalePrice
0,1,60,RL,65.0,8450,Pave,NaN,...,NaN,0,2,2008,WD,Normal,208500
1,2,20,RL,80.0,9600,Pave,NaN,...,NaN,0,5,2007,WD,Normal,181500
2,3,60,RL,68.0,11250,Pave,NaN,...,NaN,0,9,2008,WD,Normal,223500
3,4,70,RL,60.0,9550,Pave,NaN,...,NaN,0,2,2006,WD,Abnorml,140000
4,5,60,RL,84.0,14260,Pave,NaN,...,NaN,0,12,2008,WD,Normal,250000


## 2. Tính chất dữ liệu tổng quan

In [2]:
print("Số dòng:", df.shape[0])
print("Số cột:", df.shape[1])
print()
df.info(verbose=False)

So dong: 1460
So cot: 81

<class 'pandas.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Columns: 81 entries, Id to SalePrice
dtypes: float64(3), int64(35), str(43)
memory usage: 924.0 KB


In [3]:
dtype_counts = df.dtypes.value_counts()
print("Phân loại kiểu dữ liệu các cột:\n", dtype_counts)

numeric_cols = df.select_dtypes(include=[np.number]).columns.tolist()
categorical_cols = df.select_dtypes(include=['object']).columns.tolist()
print(f"\nSố biến số (numeric): {len(numeric_cols)}")
print(f"Số biến phân loại (categorical): {len(categorical_cols)}")

Phan loai kieu du lieu cac cot:
 str        43
int64      35
float64     3
Name: count, dtype: int64

So bien so (numeric): 38
So bien phan loai (categorical): 43


/tmp/ipykernel_119/3745607106.py:5: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  categorical_cols = df.select_dtypes(include=['object']).columns.tolist()


**Nhận xét:** Dữ liệu gồm 1460 quan sát, 81 cột (bao gồm `Id` và `SalePrice`).
- 38 biến dạng số (numeric: int64/float64), trong đó có cả biến định lượng thực sự (diện tích, giá) và biến có bản chất thứ tự/danh mục nhưng được lưu dưới dạng số (VD: `OverallQual`, `MSSubClass`).
- 43 biến dạng chuỗi (object/categorical), mô tả các đặc điểm danh mục của ngôi nhà (vị trí, kiểu nhà, vật liệu, chất lượng theo thang chữ...).


## 2.1. Phân biệt kiểu biến theo ý nghĩa dữ liệu

Ngoài kiểu dữ liệu mà Python đọc được (`numeric` hoặc `object`), cần nhìn vào ý nghĩa của từng biến. Theo paper của De Cock, bộ dữ liệu Ames gồm các nhóm biến **nominal, ordinal, discrete và continuous**. Vì vậy, có những biến được lưu bằng số nhưng không nên hiểu đơn giản là biến định lượng liên tục.

Ví dụ, `OverallQual` là mức đánh giá chất lượng theo thứ tự, còn `MSSubClass` là mã nhóm nhà. Ngược lại, `GrLivArea`, `TotalBsmtSF` và `LotArea` là các biến định lượng liên tục.


In [ ]:
semantic_examples = pd.DataFrame({
    'Nhóm biến': ['Định danh', 'Mục tiêu', 'Định lượng liên tục', 'Định lượng rời rạc', 'Thứ bậc', 'Phân loại'],
    'Ví dụ': [
        'Id',
        'SalePrice',
        'GrLivArea, TotalBsmtSF, LotArea',
        'BedroomAbvGr, FullBath, GarageCars',
        'OverallQual, OverallCond, KitchenQual',
        'Neighborhood, HouseStyle, SaleCondition'
    ]
})
semantic_examples


**Nhận xét bổ sung:** Như vậy, khi phân tích dữ liệu không nên chỉ dựa vào `dtype`. Cần kết hợp cả kiểu lưu trữ và ý nghĩa của biến để chọn cách phân tích phù hợp. Phần này giúp làm rõ hơn đặc điểm dữ liệu Ames được nêu trong paper.


In [4]:
df.describe().T[['count','mean','std','min','25%','50%','75%','max']].round(1).head(15)

,count,mean,std,min,25%,50%,75%,max
Id,1460.0,730.5,421.6,1.0,365.8,730.5,1095.2,1460.0
MSSubClass,1460.0,56.9,42.3,20.0,20.0,50.0,70.0,190.0
LotFrontage,1201.0,70.0,24.3,21.0,59.0,69.0,80.0,313.0
LotArea,1460.0,10516.8,9981.3,1300.0,7553.5,9478.5,11601.5,215245.0
OverallQual,1460.0,6.1,1.4,1.0,5.0,6.0,7.0,10.0
OverallCond,1460.0,5.6,1.1,1.0,5.0,5.0,6.0,9.0
YearBuilt,1460.0,1971.3,30.2,1872.0,1954.0,1973.0,2000.0,2010.0
YearRemodAdd,1460.0,1984.9,20.6,1950.0,1967.0,1994.0,2004.0,2010.0
MasVnrArea,1452.0,103.7,181.1,0.0,0.0,0.0,166.0,1600.0
BsmtFinSF1,1460.0,443.6,456.1,0.0,0.0,383.5,712.2,5644.0


## 3. Biến mục tiêu (SalePrice)

In [5]:
print(df['SalePrice'].describe())
print("\nĐộ lệch (skewness):", round(df['SalePrice'].skew(), 3))

count      1460.000000
mean     180921.195890
std       79442.502883
min       34900.000000
25%      129975.000000
50%      163000.000000
75%      214000.000000
max      755000.000000
Name: SalePrice, dtype: float64

Do lech (skewness): 1.883


**Nhận xét:** `SalePrice` dao động từ **$34,900** đến **$755,000**, trung bình khoảng **$180,921**.
Độ lệch (skewness) = **1.88** > 0 cho thấy phân phối **lệch phải khá rõ**. Phần lớn nhà có giá ở mức trung bình đến thấp, trong khi một số ít nhà có giá rất cao kéo đuôi phân phối về phía phải. Phần này sẽ được phân tích kỹ hơn ở phần đơn biến (Thành viên 2).


## 4. Danh sách nhóm cột theo nhóm ý nghĩa (theo data_description.txt)

In [6]:
with open('data_description.txt') as f:
    desc = f.read()
print(desc[:1200])

MSSubClass: Identifies the type of dwelling involved in the sale.	

        20	1-STORY 1946 & NEWER ALL STYLES
        30	1-STORY 1945 & OLDER
        40	1-STORY W/FINISHED ATTIC ALL AGES
        45	1-1/2 STORY - UNFINISHED ALL AGES
        50	1-1/2 STORY FINISHED ALL AGES
        60	2-STORY 1946 & NEWER
        70	2-STORY 1945 & OLDER
        75	2-1/2 STORY ALL AGES
        80	SPLIT OR MULTI-LEVEL
        85	SPLIT FOYER
        90	DUPLEX - ALL STYLES AND AGES
       120	1-STORY PUD (Planned Unit Development) - 1946 & NEWER
       150	1-1/2 STORY PUD - ALL AGES
       160	2-STORY PUD - 1946 & NEWER
       180	PUD - MULTILEVEL - INCL SPLIT LEV/FOYER
       190	2 FAMILY CONVERSION - ALL STYLES AND AGES

MSZoning: Identifies the general zoning classification of the sale.
		
       A	Agriculture
       C	Commercial
       FV	Floating Village Residential
       I	Industrial
       RH	Residential High Density
       RL	Residential Low Density
       RP	Residential Low Density Park 
       RM

## 5. Kết luận (Thành viên 1)

- Bài toán là **hồi quy dự đoán giá nhà** dựa trên bộ dữ liệu Ames Housing (De Cock, 2011), được dùng lại trong cuộc thi Kaggle "House Prices - Advanced Regression Techniques".
- Dữ liệu train gồm **1460 quan sát x 81 cột**, trong đó có **38 biến số** và **43 biến phân loại** theo kiểu dữ liệu mà Python đọc được.
- Biến mục tiêu **SalePrice** lệch phải khá rõ, nên cần lưu ý khi phân tích thống kê và khi xây dựng mô hình (thường có thể log-transform).
- Các phân tích chi tiết hơn về **đơn biến**, **đa biến**, **giá trị thiếu/ngoại lệ** được trình bày ở các notebook của Thành viên 2, 3, 4.
